File: keyboards/ydkb/athena75_rgb_advanced/tools/snapshot.py
#!/usr/bin/env python3
# Copyright 2026 YANG
# SPDX-License-Identifier: GPL-2.0-or-later
#
# Grab the Athena75 RGB LCD over USB (raw HID) and save it as a PNG.
#
# The firmware exposes a raw-HID command (0xFD 0x5C) that freezes core1's
# rendering and streams the shown framebuffer (RGB565, big-endian) in 27-byte
# chunks. This host tool pulls every chunk, reassembles the frame, and writes
# an image.
#
# Deps:  pip install hidapi pillow
# Usage: python snapshot.py [-o shot.png] [--vid 0x9D5B] [--pid 0x2514]

import argparse
import sys
import time

try:
    import hid
except ImportError:
    sys.exit("error: pip install hidapi")

DEFAULT_VID = 0x9D5B
DEFAULT_PID = 0x2514
USAGE_PAGE  = 0xFF60  # VIA raw-HID interface
USAGE       = 0x61

REPORT_LEN  = 32
CAP_CMD     = 0x5C
SUB_BEGIN   = 0x00
SUB_READ    = 0x01
SUB_END     = 0x02


def find_path(vid, pid):
    """Pick the VIA raw-HID interface (usage page 0xFF60 / usage 0x61)."""
    for d in hid.enumerate(vid, pid):
        if d.get("usage_page") == USAGE_PAGE and d.get("usage") == USAGE:
            return d["path"]
    # Fallback: first matching device (some platforms don't report usage).
    devs = hid.enumerate(vid, pid)
    return devs[0]["path"] if devs else None


def xfer(dev, payload, timeout_ms=1000):
    """Send one 32-byte report (report id 0) and return the 32-byte reply."""
    buf = bytearray(REPORT_LEN)
    buf[: len(payload)] = bytes(payload)
    dev.write(b"\x00" + bytes(buf))  # leading 0 = report id
    deadline = time.time() + timeout_ms / 1000.0
    while time.time() < deadline:
        rep = dev.read(REPORT_LEN, timeout_ms=50)
        if rep:
            return bytes(rep)
    raise TimeoutError("no reply from keyboard")


def capture(dev):
    # BEGIN: freeze + metadata.
    r = xfer(dev, [0xFD, CAP_CMD, SUB_BEGIN])
    if r[0] != 0xFD or r[1] != CAP_CMD:
        raise RuntimeError(f"unexpected reply header: {r[:3].hex()}")
    w = (r[3] << 8) | r[4]
    h = (r[5] << 8) | r[6]
    fmt = r[7]
    total = (r[8] << 24) | (r[9] << 16) | (r[10] << 8) | r[11]
    chunk = r[12]
    if total == 0:
        raise RuntimeError("panel reported 0 bytes (LCD off?)")
    if fmt != 2:
        raise RuntimeError(f"unsupported pixel format {fmt}")
    print(f">> frame {w}x{h}, {total} bytes, {chunk}B/chunk")

    nchunks = (total + chunk - 1) // chunk
    frame = bytearray(total)
    try:
        for idx in range(nchunks):
            r = xfer(dev, [0xFD, CAP_CMD, SUB_READ, (idx >> 8) & 0xFF, idx & 0xFF])
            # resync check on the echoed index
            if r[2] != SUB_READ or ((r[3] << 8) | r[4]) != idx:
                raise RuntimeError(f"chunk {idx}: bad reply {r[:5].hex()}")
            off = idx * chunk
            n = min(chunk, total - off)
            frame[off : off + n] = r[5 : 5 + n]
            if idx % 128 == 0 or idx == nchunks - 1:
                print(f"\r>> {idx + 1}/{nchunks}", end="", flush=True)
        print()
    finally:
        try:
            xfer(dev, [0xFD, CAP_CMD, SUB_END])
        except Exception:
            pass
    return w, h, bytes(frame)


def rgb565_be_to_rgb888(w, h, data):
    out = bytearray(w * h * 3)
    for i in range(w * h):
        px = (data[2 * i] << 8) | data[2 * i + 1]  # big-endian pair
        r5 = (px >> 11) & 0x1F
        g6 = (px >> 5) & 0x3F
        b5 = px & 0x1F
        j = 3 * i
        out[j] = (r5 << 3) | (r5 >> 2)
        out[j + 1] = (g6 << 2) | (g6 >> 4)
        out[j + 2] = (b5 << 3) | (b5 >> 2)
    return bytes(out)


def save(w, h, frame, path):
    rgb = rgb565_be_to_rgb888(w, h, frame)
    try:
        from PIL import Image

        Image.frombytes("RGB", (w, h), rgb).save(path)
    except ImportError:
        # No Pillow: fall back to a binary PPM (viewable by most tools).
        if not path.lower().endswith(".ppm"):
            path += ".ppm"
        with open(path, "wb") as f:
            f.write(f"P6\n{w} {h}\n255\n".encode())
            f.write(rgb)
    print(f">> saved {path}")


def main():
    ap = argparse.ArgumentParser(description="Screenshot the Athena75 RGB LCD over USB.")
    ap.add_argument("-o", "--out", default="lcd_shot.png", help="output image (default lcd_shot.png)")
    ap.add_argument("--vid", type=lambda s: int(s, 0), default=DEFAULT_VID)
    ap.add_argument("--pid", type=lambda s: int(s, 0), default=DEFAULT_PID)
    args = ap.parse_args()

    path = find_path(args.vid, args.pid)
    if not path:
        sys.exit(f"error: device {args.vid:#06x}:{args.pid:#06x} not found")

    dev = hid.device()
    dev.open_path(path)
    try:
        w, h, frame = capture(dev)
        save(w, h, frame, args.out)
    finally:
        dev.close()


if __name__ == "__main__":
    main()

File: keyboards/ydkb/athena75_rgb_advanced/tools/host/snapshot.c
// Copyright 2026 YANG
// SPDX-License-Identifier: GPL-2.0-or-later
//
// snapshot: grab the Athena75 RGB LCD over USB (raw-HID 0xFD 0x5C) and save a PNG.
// The firmware freezes core1 and streams the shown framebuffer (RGB565, big-
// endian) in 27-byte chunks; this reassembles the frame and writes an RGB PNG.
// Native only (Windows SetupAPI+hid.dll, macOS IOKit); PNG via the bundled writer.
//
// Usage:  host_tool snapshot [-o shot.png]

#include "cmds.h"
#include "hid.h"
#include "png.h"
#include "proto.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Send one 32-byte report and read the reply, retrying until timeout.
static int xfer(hid_dev *d, const uint8_t *payload, int plen, uint8_t *reply, int timeout_ms) {
    uint8_t buf[ATHENA_REPORT_LEN] = {0};
    if (plen > ATHENA_REPORT_LEN) plen = ATHENA_REPORT_LEN;
    memcpy(buf, payload, plen);
    if (hid_write(d, buf) != 0) return -1;
    int waited = 0;
    while (waited < timeout_ms) {
        int r = hid_read(d, reply, 100);
        if (r == 1) return 0;
        if (r < 0) return -1;
        waited += 100;
    }
    return -1; // timeout
}

int cmd_snapshot(int argc, char **argv) {
    const char *out = "lcd_shot.png";
    for (int i = 1; i < argc; i++) {
        if ((!strcmp(argv[i], "-o") || !strcmp(argv[i], "--out")) && i + 1 < argc) out = argv[++i];
        else { printf("unknown arg: %s\n", argv[i]); return 2; }
    }

    hid_dev *d = hid_open(ATHENA_VID, ATHENA_PID, ATHENA_USAGE_PAGE, ATHENA_USAGE);
    if (!d) { printf("error: device %04x:%04x not found\n", ATHENA_VID, ATHENA_PID); return 1; }

    uint8_t rep[ATHENA_REPORT_LEN];

    // BEGIN: freeze + metadata.
    uint8_t begin[] = {ATHENA_CMD, ATHENA_CAP_CMD, ATHENA_CAP_BEGIN};
    if (xfer(d, begin, sizeof begin, rep, 1000) != 0 || rep[0] != ATHENA_CMD || rep[1] != ATHENA_CAP_CMD) {
        printf("error: no/invalid reply from keyboard\n"); hid_close(d); return 1;
    }
    int w = (rep[3] << 8) | rep[4];
    int h = (rep[5] << 8) | rep[6];
    int fmt = rep[7];
    uint32_t total = ((uint32_t)rep[8] << 24) | ((uint32_t)rep[9] << 16) |
                     ((uint32_t)rep[10] << 8) | rep[11];
    int chunk = rep[12];
    if (total == 0)      { printf("error: panel reported 0 bytes (LCD off?)\n"); hid_close(d); return 1; }
    if (fmt != ATHENA_CAP_FMT_RGB565) { printf("error: unsupported pixel format %d\n", fmt); hid_close(d); return 1; }
    if (chunk <= 0 || chunk > ATHENA_CAP_CHUNK) chunk = ATHENA_CAP_CHUNK;
    printf(">> frame %dx%d, %u bytes, %dB/chunk\n", w, h, total, chunk);

    uint8_t *frame = (uint8_t *)malloc(total);
    if (!frame) { printf("error: out of memory\n"); hid_close(d); return 1; }

    uint32_t nchunks = (total + (uint32_t)chunk - 1) / (uint32_t)chunk;
    int ok = 1;
    for (uint32_t idx = 0; idx < nchunks; idx++) {
        uint8_t req[] = {ATHENA_CMD, ATHENA_CAP_CMD, ATHENA_CAP_READ,
                         (uint8_t)((idx >> 8) & 0xFF), (uint8_t)(idx & 0xFF)};
        if (xfer(d, req, sizeof req, rep, 1000) != 0 ||
            rep[2] != ATHENA_CAP_READ || (uint32_t)((rep[3] << 8) | rep[4]) != idx) {
            printf("\nerror: chunk %u: bad/no reply\n", idx); ok = 0; break;
        }
        uint32_t off = idx * (uint32_t)chunk;
        uint32_t n = total - off;
        if (n > (uint32_t)chunk) n = (uint32_t)chunk;
        memcpy(frame + off, rep + 5, n);
        if (idx % 128 == 0 || idx == nchunks - 1) { printf("\r>> %u/%u", idx + 1, nchunks); fflush(stdout); }
    }
    printf("\n");

    uint8_t endc[] = {ATHENA_CMD, ATHENA_CAP_CMD, ATHENA_CAP_END};
    xfer(d, endc, sizeof endc, rep, 500); // best effort: unfreeze core1
    hid_close(d);

    if (!ok) { free(frame); return 1; }

    // RGB565 big-endian -> RGB888.
    uint8_t *rgb = (uint8_t *)malloc((size_t)w * h * 3);
    if (!rgb) { free(frame); printf("error: out of memory\n"); return 1; }
    for (int i = 0; i < w * h; i++) {
        uint16_t px = ((uint16_t)frame[2 * i] << 8) | frame[2 * i + 1];
        int r5 = (px >> 11) & 0x1F, g6 = (px >> 5) & 0x3F, b5 = px & 0x1F;
        rgb[3 * i + 0] = (uint8_t)((r5 << 3) | (r5 >> 2));
        rgb[3 * i + 1] = (uint8_t)((g6 << 2) | (g6 >> 4));
        rgb[3 * i + 2] = (uint8_t)((b5 << 3) | (b5 >> 2));
    }
    free(frame);

    int rc = png_write_rgb(out, rgb, w, h);
    free(rgb);
    if (rc != 0) { printf("error: failed to write %s\n", out); return 1; }
    printf(">> saved %s\n", out);
    return 0;
}
